### PR TITLE
Fixes #5376

### LINKED ISSUE #5376: fetchGit failes with non default ssh port
Since https://github.com/NixOS/nix/pull/4978 was merged, fetchGit do not respect a ssh port.

**Steps To Reproduce**

With current Nix 2.4pre20211006_53e4794:
```
nix-repl> builtins.fetchGit { url = "ssh://git@github.com:22/NixOS/nix.git"; }
fetching Git repository 'ssh://git@github.com/22/NixOS/nix.git'fatal: remote error:
  22/NixOS/nix is not a valid repository name
  Visit https://support.github.com/ for help
error: program 'git' failed with exit code 128
```

With old Nix 2.4pre20210922_bcd73eb:
```
nix-repl> builtins.fetchGit { url = "ssh://git@github.com:22/NixOS/nix.git"; }
{ lastModified = 1634049780; lastModifiedDate = "20211012144300"; narHash = "sha256-NYjt6jTZyt35hJJfjQsVPB4w7iaN/z40Vzp0k1q1cKA="; outPath = "/nix/store/k302z97m35fl3cnycrgq814kfb490f48-source"; rev = "5176b072ed55d1528cb36e0c865e4fcffbc6a7a1"; revCount = 10824; shortRev = "5176b07"; submodules = true; }
```

**Expected behavior**

fetchGit shall respect the given ssh port as it does for Nix 2.4pre20210922_bcd73eb.

### PR DESCRIPTION
**Pull request by ck3d**
scp uri can not contain a slash before @

Fixes #5376

### CODE REVIEW

**Comment by regnat** [comment]
> scp uri can not contain a slash before @

I don’t remember exactly the details here, but unless I miss something, the url *will* contain a slash if it starts with a scheme (`ssh://foo@bar.com`), so the new regex should break, right?

File: src/libexpr/primops/fetchTree.cc [hunk 1]
```diff
@@ -74,7 +74,7 @@ std::string fixURI(std::string uri, EvalState & state, const std::string & defau
 
 std::string fixURIForGit(std::string uri, EvalState & state)
 {
-    static std::regex scp_uri("([^/].*)@(.*):(.*)");
+    static std::regex scp_uri("^([^/]*)@(.*):(.*)");
```
Comment (edolstra) [inline src/libexpr/primops/fetchTree.cc:77]: Note that a ^ at the start should be unnecessary since `regex_match` already must match the entire input.

Comment (ck3d) [reply src/libexpr/primops/fetchTree.cc]: Thanks, works.

**Comment by ck3d** [comment]
Yes, the prefix ssh:// can be used to distinguish between scp and ssh.

**Comment by ck3d** [comment]
I hope this fix can be also merged into 2.4 soon, since I have to specify the port in my projects.

**Comment by ck3d** [comment]
Please have a look before releasing 2.4.1 ... This PR fixes the mentioned regression which was introduced with 2.4.

What can I to get progress in this PR?

**Comment by ck3d** [comment]
I simplified the patch by keeping the regex untouched.

**Comment by ck3d** [comment]
PR #5580 was merged.

**Pull request closed without merging**